fms_core/fzy.py

```diff
@@ -0,0 +1,207 @@
+                elif j:  # i > 0 and j > 0*/
+                    score = max([
+                        M[i - 1][j - 1] + match_bonus[j],
+
+                        # consecutive match, doesn't stack with match_bonus
+                        D[i - 1][j - 1] + SCORE_MATCH_CONSECUTIVE])
+
+                D[i][j] = score
+                M[i][j] = prev_score = max([score, prev_score + gap_score])
+            else:
+                D[i][j] = SCORE_MIN
+                M[i][j] = prev_score = prev_score + gap_score
+
+
+def score(needle, haystack):
+    n = len(needle)
+    m = len(haystack)
+
+    if not n or not m:
+        return SCORE_MIN
+
+    #  if n == m:
+        #  #
+        #  # Since this method can only be called with a haystack which
+        #  # matches needle. If len(the)s of the strings are equal the
+        #  # strings themselves must also be equal (ignoring case).
+        #  #
+        #  return SCORE_MAX
+
+    if m > 1024:
+        #
+        # Unreasonably large candidate: return no score
+        # If it is a valid match it will still be returned, it will
+        # just be ranked below any reasonably sized candidates
+        #
+        return SCORE_MIN
+
+    D = [0] * n
+    M = [0] * n
+
+    compute(needle, haystack, D, M)
+
+    return M[n - 1][m - 1]
+
+
+def positions(needle, haystack):
+    n = len(needle)
+    m = len(haystack)
+
+    positions = []
+
+    if not n or not m:
+        return positions
+
+    if n == m:
+        for i in range(0, n):
+            positions[i] = i
+        return positions
+
+    if m > 1024:
+        return positions
+
+    D = [0] * n
+    M = [0] * n
+
+    compute(needle, haystack, D, M)
+
+    # backtrack to find the positions of optimal matching
+    match_required = False
+
+    i = n - 1
+    j = m - 1
+    while i >= 0:
+        while j >= 0:
+            #
+            # There may be multiple paths which result in
+            # the optimal weight.
+            #
+            # For simplicity, we will pick the first one
+            # we encounter, the latest in the candidate
+            # string.
+            #
+            if D[i][j] != SCORE_MIN and (match_required or D[i][j] == M[i][j]):
+                # If this score was determined using
+                # SCORE_MATCH_CONSECUTIVE, the
+                # previous character MUST be a match
+
+                match_required = i and j and \
+                    M[i][j] == D[i - 1][j - 1] + SCORE_MATCH_CONSECUTIVE
+                positions[i] = j
+                j -= 1
+                break
+            j -= 1
+
+        i -= 1
+
+    return positions
+
+
+def hasMatch(needle, haystack):
+    needle = needle.lower()
+    haystack = haystack.lower()
+    length = len(needle)
+    i = 0
+    j = 0
+    while i < length:
+        j = haystack.index(needle[i], j) + 1
+        if j == 0:
+            return False
+        i += 1
+
+    return True
+
+#  export {
```

Comment:
delete

Reply:
(if you want to leave something semantic about the public parts of the module, use `__all__`)